---
title: "Post-processing"
author: "Alyssa Imbert and Etienne A. Thévenot (ProMetIS consortium)"
date: "`r doc_date()`"

vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: "ProMetIS.bib"
output:
  BiocStyle::html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: false
---

![](figures/prometis_logo.png)

```{r options, include=FALSE}
knitr::opts_chunk$set(fig.width = 8,
                      fig.height = 8,
                      fig.path = 'figures/temp/2_post_processed/',
                      warning = FALSE)
```

# Mice IDs

```{r mice_id}
mice_id.df <- read.table(file.path(ProMetIS::processed_dir.c(),
                                   "mice_id.tsv"),
                         row.names = 1,
                         header = TRUE, sep = "\t")
mice_id.vc <- rownames(mice_id.df)
mice_num.vc <- substr(mice_id.vc, 2, 4)
stopifnot(identical(as.integer(mice_num.vc),
                    sort(as.integer(mice_num.vc))))
mice.ls <- list(id.df = mice_id.df,
                id.vc = mice_id.vc,
                num.vc = mice_num.vc)
```

# Preclinical

## Loading

```{r preclinical_loading}
preclinic_setname.c <- "preclinical"

# table from PHENOMIN (47 * 1098)
preclinic_raw.df <- read.table(file.path(ProMetIS::processed_dir.c(),
                                         preclinic_setname.c,
                                         "phenomin.tsv"),
                               check.names = FALSE,
                               #comment.char = "",
                               header = TRUE, quote = "", sep = "\t")
preclinic_proc_names.vc <- colnames(preclinic_raw.df)
colnames(preclinic_raw.df) <- make.names(colnames(preclinic_raw.df), unique = TRUE)

# checking sample order
stopifnot(identical(preclinic_raw.df[, "mouse_nb"],
                    sort(preclinic_raw.df[, "mouse_nb"])))
```

## sampleMetadata

```{r preclinical_sampleMetadata}
# restricting to the 42 samples analyzed in ProMetIS

preclinic_raw.df <- preclinic_raw.df[as.character(preclinic_raw.df[, "mouse_nb"]) %in% mice_num.vc, ]
stopifnot(identical(as.character(preclinic_raw.df[, "mouse_nb"]), mice_num.vc))
rownames(preclinic_raw.df) <- mice_id.vc

# sampleMetadata (42 * 1098)

preclinic_pda.df <- preclinic_raw.df

# # labeling the 'initial' information (to be kept at the end of the
# # sampleMetadata and variableMetadata tables)
# 
# supp.vi <- which(!(colnames(preclinic_pda.df) %in% ProMetIS:::.varLabels_first.ls()[["preclinical"]]))
# stopifnot(length(supp.vi) == 1091)
# colnames(preclinic_pda.df)[supp.vi] <- paste0("supp_",
#                                            colnames(preclinic_pda.df)[supp.vi])
```

## dataMatrix

```{r preclinical_dataMatrix}
# initial dataMatrix (42 * 1091)

variable.vi <-  which(!(colnames(preclinic_raw.df) %in% c(ProMetIS:::.sample_metadata_select("preclinical"), "mouse_id", "genotype", "project"))) 
stopifnot(length(variable.vi) == 1091)
preclinic_dat.df <- preclinic_raw.df[, variable.vi]
dat_proc_names.vc <- preclinic_proc_names.vc[variable.vi]

rm(preclinic_raw.df)
rm(preclinic_proc_names.vc)

## discarding variables which cannot be converted to numerics
### removing 1 variable with all values identical except 1 (for "W627m")
### note that all 'Test.de.Tolérance.au.Glucose' tests have a 'NA' value for this mouse only

discard_imagerie.vl <- grepl("Imagerie.du.squelette.par.rayons.X.Digits.integrity",
                             colnames(preclinic_dat.df))
stopifnot(sum(discard_imagerie.vl) == 1)

### removing date of sacrifice
discard_sacrifice.vl <- grepl("Date.of.sacrifice", colnames(preclinic_dat.df))
stopifnot(sum(discard_sacrifice.vl) == 2)

### removing 471 columns with NA or "" values only
discard_navoid.vl <- apply(preclinic_dat.df, 2,
                           function(y) {
                             unique.vc <- unique(as.character(y))
                             unique.vc <- unique.vc[!is.na(unique.vc) & (unique.vc != "")]
                             length(unique.vc) < 2
                           })
stopifnot(sum(discard_navoid.vl) == 478)

discard_feat.vl <- discard_imagerie.vl |
  discard_sacrifice.vl |
  discard_navoid.vl
stopifnot(sum(discard_feat.vl) == 481)

preclinic_dat.df <- preclinic_dat.df[, !discard_feat.vl]
dat_proc_names.vc <- dat_proc_names.vc[!discard_feat.vl]

## keeping numerics and factors with nlevels > 2

preclinic_class.vc <- sapply(preclinic_dat.df, data.class)
for (preclinic_char.i in which(preclinic_class.vc == "character"))
  preclinic_dat.df[, preclinic_char.i] <- factor(preclinic_dat.df[, preclinic_char.i])
preclinic_class.vc <- sapply(preclinic_dat.df, data.class)
stopifnot(identical(unique(preclinic_class.vc), c("numeric", "factor")))

## selecting numerics (678)
keep_numeric.vl <- preclinic_class.vc == "numeric"
stopifnot(sum(keep_numeric.vl) == 571)

## selecting factors with 2 levels (45)
keep_factor2.vl <- logical(ncol(preclinic_dat.df))
for (j in which(preclinic_class.vc == "factor")) {
  preclinic_dat.f <- preclinic_dat.df[, j]
  if (nlevels(preclinic_dat.f) == 2) {
    preclinic_dat.df[, j] <- as.integer(preclinic_dat.df[, j])
    keep_factor2.vl[j] <- TRUE
  }
}
stopifnot(sum(keep_factor2.vl) == 21)

## aggregating selections (725)
keep_aggreg.vl <- keep_numeric.vl |
  keep_factor2.vl
stopifnot(sum(keep_aggreg.vl) == 592)
preclinic_dat.mn <- as.matrix(preclinic_dat.df[, keep_aggreg.vl])
dat_proc_names.vc <- dat_proc_names.vc[keep_aggreg.vl]

## removing variables with > 20% NA (including 100% NA in females)
## or with variance = 0
filtered.eset <- phenomis::filtering(Biobase::ExpressionSet(assayData = t(preclinic_dat.mn),
                                                            phenoData = Biobase::AnnotatedDataFrame(data.frame(row.names = rownames(preclinic_dat.mn),
                                                                                                               gene = substr(rownames(preclinic_dat.mn), 1, 1)))),
                                     class.c = "gene",
                                     dims.vc = "features")
keep_notna_zerovar.vl <- colnames(preclinic_dat.mn) %in% Biobase::featureNames(filtered.eset)

# keep_notna_zerovar.vl <- ProMetIS:::.filter_na_zerovar(preclinic_dat.mn,
#                                                        na_thresh.n = 0.2)

stopifnot(sum(keep_notna_zerovar.vl) == 238)
preclinic_dat.mn <- preclinic_dat.mn[, keep_notna_zerovar.vl]
dat_proc_names.vc <- dat_proc_names.vc[keep_notna_zerovar.vl]

stopifnot(identical(dim(preclinic_dat.mn), as.integer(c(42, 238))))
```


## variableMetadata

```{r preclinical_variableMetadata}
# variableMetadata (238 * 3)
category.vc <- sapply(colnames(preclinic_dat.mn),
                      function(name.c)
                        unlist(strsplit(name.c,
                                        split = ".",
                                        fixed = TRUE))[1])
category.vc <- gsub("Auditory", "Auditory and PPI",
                    gsub("Clinical", "Clinical chemistry",
                         gsub("Grip", "Grip test",
                              gsub("Hot", "Hot plate",
                                   gsub("Modified", "SHIRPA",
                                        gsub("Open", "Open field",
                                             gsub("Pavlovian", "Pavlovian fear cond.",
                                                  gsub("Shock", "Shock threshold",
                                                       gsub("Test", "Glucose tolerance test",
                                                            gsub("Ultrafocus", "Body composition",
                                                                 category.vc))))))))))
full.vc <- sapply(colnames(preclinic_dat.mn),
                  function(name.c) {
                    name.vc <- unlist(strsplit(name.c,
                                               split = ".",
                                               fixed = TRUE))
                    if (length(name.vc) == 1) {
                      return("")
                    } else {
                      name.vc <- name.vc[-1]
                      name.vc <- name.vc[name.vc != ""]
                      return(paste(name.vc, collapse = "_"))
                    }
                  })
preclinic_fda.df <- data.frame(row.names = colnames(preclinic_dat.mn),
                               measurement = dat_proc_names.vc,
                               category = category.vc,
                               stringsAsFactors = FALSE)
stopifnot(identical(dim(preclinic_fda.df), as.integer(c(238, 2))))
```

## ExpressionSet

```{r preclinical_ExpressionSet}
# converting to ExpressionSet
preclinical.eset <- Biobase::ExpressionSet(assayData = t(preclinic_dat.mn),
                                           phenoData = Biobase::AnnotatedDataFrame(data = preclinic_pda.df),
                                           featureData = Biobase::AnnotatedDataFrame(data = preclinic_fda.df),
                                           experimentData = Biobase::MIAME(title = preclinic_setname.c))
stopifnot(methods::validObject(preclinical.eset))

preclinical.eset <- ProMetIS:::metadata_select(preclinical.eset, step.c = "2_post_processed")[["preclinical"]]

print(preclinical.eset)
```

## Deleting the Ca outlier measurements

```{r preclinical_outliers}
preclinical.eset <- preclinical.eset[!grepl("Ca", Biobase::featureNames(preclinical.eset)), ]
```

## Saving (not run)

```{r preclinical_saving}
# an 'id' extra column has been automatically created in the sampleMetadata by
# 'MultiDataSet' and must be removed before saving
pdata.df <- Biobase::pData(preclinical.eset)
pdata.df[, "id"] <- NULL
Biobase::pData(preclinical.eset) <- pdata.df

phenomis::writing(preclinical.eset,
                  file.path(gsub(ProMetIS::data_dir.c(),
                                 "../../ProMetIS/inst/extdata",
                                 ProMetIS::post_processed_dir.c()),
                            preclinic_setname.c),
                  overwrite.l = TRUE)
```


# Proteomics

## Set names and input files

```{r proteo_files}
proteo_files.vc <- vapply(ProMetIS::proteo_sets.vc(),
                          function(set.c) {
                            file.c <- list.files(file.path(ProMetIS::processed_dir.c(), set.c),
                                                 pattern = ".xlsx", full.names = TRUE)
                            stopifnot(length(file.c) == 1)
                            file.c
                          }, FUN.VALUE = character(1))

proteo.mset <- MultiDataSet::createMultiDataSet()
```

## MultiDataSet containing the data matrices

```{r proteo_dataMatrix}
for (set.c in ProMetIS::proteo_sets.vc()) {
  
  # dataMatrix
  data.df <- as.data.frame(readxl::read_excel(proteo_files.vc[set.c],
                                              sheet = 1),
                           stringsAsFactors = FALSE)
  rownames(data.df) <- data.df[, 1]
  data.df[, 1] <- NULL
  data.mn <- as.matrix(data.df)
  rm(data.df)
  mode(data.mn) <- "numeric"
  
  eset <- Biobase::ExpressionSet(assayData = data.mn,
                                 experimentData = Biobase::MIAME(title = set.c))
  stopifnot(methods::validObject(eset))
  
  proteo.mset <- MultiDataSet::add_eset(proteo.mset,
                                        eset,
                                        dataset.type = set.c,
                                        GRanges = NA,
                                        overwrite = TRUE,
                                        warnings = FALSE)
}
```

## sampleMetadata

```{r proteo_sampleMetadat}
for (set.c in ProMetIS::proteo_sets.vc()) {
  
  eset <- proteo.mset[[set.c]]
 
  proteo_pda.df <- as.data.frame(readxl::read_excel(proteo_files.vc[set.c],
                                                    sheet = 2),
                                 stringsAsFactors = FALSE)
  rownames(proteo_pda.df) <- gsub(".", "_",
                                  proteo_pda.df[, 1], fixed = TRUE)
  sample_names.vc <- Biobase::sampleNames(eset)
  if (set.c == "proteomics_liver") {
    stopifnot(identical(sort(rownames(proteo_pda.df)),
                        sort(sample_names.vc)))
    proteo_pda.df <- proteo_pda.df[sample_names.vc, ]
  }
  stopifnot(identical(rownames(proteo_pda.df),
                      sample_names.vc))
  
  Biobase::pData(eset) <- proteo_pda.df
  stopifnot(methods::validObject(eset))
  
  proteo.mset <- MultiDataSet::add_eset(proteo.mset,
                                        eset,
                                        dataset.type = set.c,
                                        GRanges = NA,
                                        overwrite = TRUE,
                                        warnings = FALSE)
}
```

## variableMetadata

```{r proteo_variableMetadata}
for (set.c in ProMetIS::proteo_sets.vc()) {
  
  eset <- proteo.mset[[set.c]]

  proteo_fda.df <- as.data.frame(readxl::read_excel(proteo_files.vc[set.c],
                                                    sheet = 3),
                                 stringsAsFactors = FALSE)
  rownames(proteo_fda.df) <- proteo_fda.df[, 1]
  stopifnot(identical(rownames(proteo_fda.df),
                      Biobase::featureNames(eset)))
  
  proteo_fda.df[, "uniprot_id"] <- sapply(proteo_fda.df[, "accession"],
                                          function(access.c)
                                            unlist(strsplit(access.c,
                                                            split = "|",
                                                            fixed = TRUE))[2],
                                          USE.NAMES = FALSE)
  
  Biobase::fData(eset) <- proteo_fda.df
  stopifnot(methods::validObject(eset))
  
  proteo.mset <- MultiDataSet::add_eset(proteo.mset,
                                        eset,
                                        dataset.type = set.c,
                                        GRanges = NA,
                                        overwrite = TRUE,
                                        warnings = FALSE)
}
```

## Renaming features and re-ordering samples

```{r proteo_reordering}
for (set.c in ProMetIS::proteo_sets.vc()) {
  
  eset <- proteo.mset[[set.c]]
  
  # changing variable IDs to names
  feat_names.vc <- paste0(sapply(Biobase::fData(eset)[, "accession"],
                                 function(access.c)
                                   unlist(strsplit(access.c, split = "|",
                                                   fixed = TRUE))[2]),
                          "_",
                          sapply(Biobase::fData(eset)[, "description"],
                                 function(desc.c) {
                                   if (!is.na(desc.c) && desc.c != "") {
                                     desc.c <- unlist(strsplit(desc.c, split = "  OS="))[1]
                                     if (nchar(desc.c) > 25)
                                       desc.c <- paste0(substr(desc.c, 1, 24), ".")
                                     return(desc.c)
                                   }
                                 }))
  stopifnot(!any(duplicated(feat_names.vc)))
  Biobase::featureNames(eset) <- feat_names.vc
  
  # discarding pools
  if (set.c == "proteomics_liver")
    eset <- eset[, !grepl("(p|P)ool", Biobase::pData(eset)[, "sample name"])]

  # re-ordering samples
  if (set.c == "proteomics_liver") {
    eset <- eset[, order(as.integer(Biobase::pData(eset)[, "sample name"]))]
  } else {
    Biobase::sampleNames(eset) <- sapply(Biobase::sampleNames(eset),
                                                 function(samp.c) unlist(strsplit(samp.c, split = "_"))[2], USE.NAMES = FALSE)
    eset <- eset[, order(as.integer(Biobase::sampleNames(eset)))]
  }
  
  # renaming samples
  if (set.c == "proteomics_liver") {
    Biobase::sampleNames(eset) <- paste0(sapply(Biobase::pData(eset)[, "Condition"],
                                                        function(cond.c)
                                                          switch(cond.c,
                                                                 mx2 = "X",
                                                                 ctrl = "W",
                                                                 lat = "L")),
                                                 Biobase::pData(eset)[, "sample name"])
  } else {
    Biobase::sampleNames(eset) <- paste0(sapply(Biobase::pData(eset)[, "Condition"],
                                                        function(cond.c)
                                                          switch(cond.c,
                                                                 Mx2 = "X",
                                                                 CONT = "W",
                                                                 LAT = "L")),
                                                 Biobase::sampleNames(eset))
  }
  
  mice_konum.vc <- substr(mice_id.vc, 1, 4)
  mice_koid.vc <- mice_id.vc
  if (set.c == "proteomics_plasma") {
    stopifnot(all(Biobase::sampleNames(eset) %in% mice_konum.vc))
    mice_koid.vc <- mice_koid.vc[mice_konum.vc %in% Biobase::sampleNames(eset)]
    mice_konum.vc <- mice_konum.vc[mice_konum.vc %in% Biobase::sampleNames(eset)]
  }
  
  stopifnot(identical(Biobase::sampleNames(eset),
                      mice_konum.vc))
  
  Biobase::sampleNames(eset) <- mice_koid.vc
  stopifnot(methods::validObject(eset))
  
  proteo.mset <- MultiDataSet::add_eset(proteo.mset,
                                        eset,
                                        dataset.type = set.c,
                                        GRanges = NA,
                                        overwrite = TRUE,
                                        warnings = FALSE)
}
```

## Adding meta data (gene, mouse_nb, sex)

```{r proteo_adding_metadata}

for (set.c in ProMetIS::proteo_sets.vc()) {
  
  eset <- proteo.mset[[set.c]]
  
  Biobase::pData(eset) <- cbind.data.frame(mice_id.df[Biobase::sampleNames(eset), ],
                                           Biobase::pData(eset),
                                           stringsAsFactors = FALSE)
  stopifnot(methods::validObject(eset))
  
  proteo.mset <- MultiDataSet::add_eset(proteo.mset,
                                        eset,
                                        dataset.type = set.c,
                                        GRanges = NA,
                                        overwrite = TRUE,
                                        warnings = FALSE)
  
}
```

## Variable metadata: including new mice ids in the column names (count, abundance, etc.)

```{r proteo_supp}
for (set.c in ProMetIS::proteo_sets.vc()) {
  
  eset <- proteo.mset[[set.c]]
  
  samp.vc <- Biobase::sampleNames(eset)
  pdata.df <- Biobase::pData(eset)
  fvar.vc <- Biobase::fvarLabels(eset)
 
  for (i in seq_along(samp.vc)) {
    samp.c <- samp.vc[i]
    samp_id.c <- gsub("abundance_", "", pdata.df[i, "id"])
    fvar_id.vi <- grep(paste0(".+", samp_id.c), fvar.vc)
    fvar.vc[fvar_id.vi] <- gsub(switch(set.c,
                                       proteomics_liver = "mgf",
                                       proteomics_plasma = samp_id.c),
                                samp.c, fvar.vc[fvar_id.vi])
  }
  
  Biobase::fvarLabels(eset) <- fvar.vc
  
  stopifnot(methods::validObject(eset))
  
  proteo.mset <- MultiDataSet::add_eset(proteo.mset,
                                        eset,
                                        dataset.type = set.c,
                                        GRanges = NA,
                                        overwrite = TRUE,
                                        warnings = FALSE)
  
}
```

## Re-ordering metadata ('supp_' columns at the end)

```{r proteo_reordering_supp}
proteo.mset <- ProMetIS:::metadata_select(proteo.mset, step.c = "2_post_processed")
```


## Saving (not run)

```{r proteo_saving}
for (set.c in ProMetIS::proteo_sets.vc()) {
  
  eset <- proteo.mset[[set.c]]
  
  # an 'id' extra column has been automatically created in the sampleMetadata by
  # 'MultiDataSet' and must be removed before saving
  pdata.df <- Biobase::pData(eset)
  pdata.df[, "id"] <- NULL
  Biobase::pData(eset) <- pdata.df
  
  phenomis::writing(eset,
                  file.path(gsub(ProMetIS::data_dir.c(),
                                 "../../ProMetIS/inst/extdata",
                                 ProMetIS::post_processed_dir.c()),
                            set.c),
                  overwrite.l = TRUE)
}

```

# Metabolomics

## Loading

```{r metabo_loading}
metabo.mset <- phenomis::reading(file.path(ProMetIS::processed_dir.c()),
                                 subsets.vc = ProMetIS::metabo_sets.vc(),
                                 report.c = "none")
```

## Normalization of the dataMatrix

The workflow consists of the following steps:

|step                |liver_c18hypersil_pos |liver_hilic_neg    |plasma_c18hypersil_pos|plasma_hilic_neg   |plasma_c18acquity_pos|plasma_c18acquity_neg|
|--------------------|----------------------|-------------------|----------------------|-------------------|---------------------|---------------------|
|0.4 <= RT <= 22     |                      |                   |                      |                   |          X          |           X         |
|blank/samp   <= 0.33|           X          |         X         |           X          |         X         |          X          |           X         |
|poolDil: cor >= 0.7 |           X          |         X         |           X          |         X         |                     |                     |
|drift loess, span=1 |          none        |        none       |          pool        |        pool       |        sample       |         sample      |
|NA <= 20% & var > 0 |           X          |         X         |           X          |         X         |          X          |           X         |
|pool CV       <= 0.3|           X          |         X         |           X          |         X         |          X          |           X         |
|poolCV/sampCV <= 1  |           X          |         X         |           X          |         X         |          X          |           X         |
|corr, dRT, dmz      |           X          |         X         |           X          |         X         |          X          |           X         |

```{r metabolomics_dataMatrix, message=FALSE, warning=FALSE}
metabo.mset <- ProMetIS:::.metabo_postprocessing(metabo.mset = metabo.mset,
                                                 drift_correct.c = "prometis")
```

## Formatting and ordering sample names

```{r metabo_format_names}
metabo.mset <- ProMetIS:::.format_metabonames(metabo.mset = metabo.mset,
                                              mice.ls = mice.ls)
```

## Log2 transformation

```{r metabo_log}
metabo.mset <- phenomis::transforming(metabo.mset, report.c = "none")
```

## Re-ordering metadata ('supp_' columns at the end)

```{r metabo_reordering}
metabo.mset <- ProMetIS:::metadata_select(metabo.mset,
                                          step.c = "2_post_processed")
```


## Saving (not run)

```{r metabo_saving}
for (set.c in ProMetIS::metabo_sets.vc()) {
  
  eset <- metabo.mset[[set.c]]
  
  # an 'id' extra column has been automatically created in the sampleMetadata by
  # 'MultiDataSet' and must be removed before saving
  pdata.df <- Biobase::pData(eset)
  pdata.df[, "id"] <- NULL
  Biobase::pData(eset) <- pdata.df
  
  phenomis::writing(eset,
                    file.path(gsub(ProMetIS::data_dir.c(),
                                   "../../ProMetIS/inst/extdata",
                                   ProMetIS::post_processed_dir.c()),
                              set.c),
                    overwrite.l = TRUE)
  
}
```

